program/src/oracle/upd_aggregate.h

```diff
@@ -205,39 +206,76 @@ static inline void upd_aggregate( pc_price_t *ptr, uint64_t slot )
+      pc_price_comp_t *iptr = &ptr->comp_[i];
+      // copy contributing price to aggregate snapshot
+      iptr->agg_ = iptr->latest_;
+      // add quote to sorted permutation array if it is valid
+      int64_t slot_diff = ( int64_t )slot - ( int64_t )( iptr->agg_.pub_slot_ );
+      int64_t price     = iptr->agg_.price_;
+      int64_t conf      = ( int64_t )( iptr->agg_.conf_ );
+      if ( iptr->agg_.status_ == PC_STATUS_TRADING &&
+           (int64_t)0 < conf && conf < price && conf <= (INT64_MAX-price) && // No overflow for INT64_MAX-price as price>0
+           slot_diff >= 0 && slot_diff <= PC_MAX_SEND_LATENCY ) {
+        vidx[numv++] = i; // FIXME: GRAFT FOR OLD PRICE MODEL
+        prcs[ nprcs++ ] = price - conf; // No overflow as 0 < conf < price
+        prcs[ nprcs++ ] = price;
+        prcs[ nprcs++ ] = price + conf; // No overflow as 0 < conf <= INT64_MAX-price
+      }
+    }
+
+    // too few valid quotes
+    ptr->num_qt_ = numv; // FIXME: TEMPORARY GRAFT (ALL RETURN PATHS SET NUM_QT TO SOMETHING SENSIBLE)
+    if ( numv == 0 || numv < ptr->min_pub_ ) {
+      ptr->agg_.status_ = PC_STATUS_UNKNOWN;
+      return;
     }
-  }
 
-  uint32_t nprcs = 0;
-  int64_t prcs[ PC_COMP_SIZE * 2 ];
-  for ( uint32_t i = 0; i < numv; ++i ) {
-    pc_price_comp_t const *iptr = &ptr->comp_[ vidx[ i ] ];
-    int64_t const price = iptr->agg_.price_;
-    int64_t const conf = ( int64_t )( iptr->agg_.conf_ );
-    prcs[ nprcs++ ] = price - conf;
-    prcs[ nprcs++ ] = price + conf;
+    // evaluate the model to get the p25/p50/p75 prices
+    // note: numv>0 and nprcs = 3*numv at this point
+    int64_t agg_p25;
+    int64_t agg_p75;
+    int64_t scratch[ PC_COMP_SIZE * 3 ]; // ~0.75KiB for current PC_COMP_SIZE (FIXME: DOUBLE CHECK THIS FITS INTO STACK FRAME LIMIT)
+    price_model_core( (uint64_t)nprcs, prcs, &agg_p25, &agg_price, &agg_p75, scratch );
+
+    // get the left and right confidences
+    // note that as valid quotes have positive prices currently and
+    // agg_p25, agg_price, agg_p75 are ordered, this calculation can't
+    // overflow / underflow
+    int64_t agg_conf_left  = agg_price - agg_p25;
+    int64_t agg_conf_right = agg_p75 - agg_price;
+
+    // use the larger of the left and right confidences
+    agg_conf = agg_conf_right > agg_conf_left ? agg_conf_right : agg_conf_left;
+
+    // if the confidences end up at zero, we abort
+    // this is paranoia as it is currently not possible when nprcs>2 and
+    // positive confidences given the current pricing model
+    if( agg_conf <= (int64_t)0 ) {
+      ptr->agg_.status_ = PC_STATUS_UNKNOWN;
+      return;
+    }
   }
-  qsort_int64( prcs, nprcs );
+
+  // FIXME: MOST OF THE REST OF THE CODE HERE CAN PROBABLY BE REMOVED /
+  // CONDENSED AND/OR MADE MORE EFFICIENT.  AND IT PROBABLY WILL
+  // REPLACED SOON BY THE VWAP MODEL UNDER DEVELOPMENT.  IT IS KEPT HERE
+  // TO KEEP THE CURRENT VWAP CALCULATION AS CLOSE AS POSSIBLE TO WHAT
+  // HOW IT CURRENTLY WORKS.
```

Comment:
let's actually do this comment and delete the old price model from below. we can keep the old twap computation and feed the new aggregate price / confidence into it. I believe most of the code below is dead.